cdap-cli/src/main/java/co/cask/cdap/cli/command/app/BaseBatchCommand.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.cli.command.app;
+
+import co.cask.cdap.cli.ArgumentName;
+import co.cask.cdap.cli.CLIConfig;
+import co.cask.cdap.cli.ElementType;
+import co.cask.cdap.cli.english.Article;
+import co.cask.cdap.cli.english.Fragment;
+import co.cask.cdap.cli.util.AbstractAuthCommand;
+import co.cask.cdap.client.ApplicationClient;
+import co.cask.cdap.common.ApplicationNotFoundException;
+import co.cask.cdap.common.UnauthorizedException;
+import co.cask.cdap.proto.BatchProgram;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramRecord;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.common.cli.Arguments;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for commands that work on multiple programs in an application.
+ *
+ * @param <T> the type of input object for the batch request
+ */
+public abstract class BaseBatchCommand<T extends BatchProgram> extends AbstractAuthCommand {
+  private final ApplicationClient appClient;
+
+  protected BaseBatchCommand(ApplicationClient appClient, CLIConfig cliConfig) {
+    super(cliConfig);
+    this.appClient = appClient;
+  }
+
+  @Override
+  public void perform(Arguments arguments, PrintStream output) throws Exception {
+    Args<T> args = readArgs(arguments);
+    if (args.programs.isEmpty()) {
+      output.printf(String.format("application '%s' contains no programs of type '%s'",
```

Comment:
ah, I see, ok:
`appClient.listProgramsByType(appId);`
I assumed that the `args` was constructed entirely and solely based off of the user arguments.